dipy/reconst/ivim.py

```diff
@@ -0,0 +1,585 @@
+
+    def __getitem__(self, index):
+        model_params = self.model_params
+        N = model_params.ndim
+        if type(index) is not tuple:
+            index = (index,)
+        elif len(index) >= model_params.ndim:
+            raise IndexError("IndexError: invalid index")
+        index = index + (slice(None),) * (N - len(index))
+        return type(self)(self.model, model_params[index])
+
+    @property
+    def S0_predicted(self):
+        return self.model_params[..., 0]
+
+    @property
+    def perfusion_fraction(self):
+        return self.model_params[..., 1]
+
+    @property
+    def D_star(self):
+        return self.model_params[..., 2]
+
+    @property
+    def D(self):
+        return self.model_params[..., 3]
+
+    @property
+    def shape(self):
+        return self.model_params.shape[:-1]
+
+    def predict(self, gtab, S0=1.):
+        """Given a model fit, predict the signal.
+
+        Parameters
+        ----------
+        gtab : GradientTable class instance
+               Gradient directions and bvalues
+
+        S0 : float
+            S0 value here is not necessary and will not be used to predict the
+            signal. It has been added to conform to the structure of the
+            predict method in multi_voxel which requires a keyword argument S0.
+
+        Returns
+        -------
+        signal : array
+            The signal values predicted for this model using its parameters.
+
+        """
+        return ivim_prediction(self.model_params, gtab)
+
+    def fill_na(self, bounds=((0., 0., 0., 0), (np.inf, 0.30, 0.01, 0.001)),
+                fill=(np.nan, np.nan, np.nan, np.nan)):
+        """
+        Function to fill nan values for parameters which are not within the bounds.
+        This function will check if `f `is within bounds and set D_star and f to nan
+        if `f` is out of bounds.
+
+        Parameters
+        ----------
+        bounds : tuple
+            A tuple of two elements specifying the bounds. To check only for a
+            particular parameter (say `f`) set the other bounds to
+            (-np.inf, np.inf).
+            default : ((0., 0., 0., 0),(np.inf, 0.30, 0.01, 0.001))
+
+        fill : tuple
+            A tuple which specifies what should be replaced for the parameters
+            if bounds is violated. If a particular parameter should not be
+            changed then set the value as True. For example, in order to get
+            the S0 and D values and fill `nan` for others use
+            `fill=(True, np.nan, np.nan, True). Similarly, to set `f` and
+            `D_star` as 0. and leave S0 and D unchanged, use
+            `fill=(True, 0., 0., True)`. Using `False` for any fill will set
+            it to 0.
+            default : (np.nan, np.nan, np.nan, np.nan)
+
+        Returns
+        -------
+        IvimFit : IvimFit object
+        """
+        params = self.model_params
+        bounds_violated = ~(np.all(params[..., :] >= bounds[
```

Comment:
Also, is this working as you intend? I think the `np.all` call is collapsing the whole array down to a single `True` or `False`. It then would replace all the parameters in the array rather that each element individually. So unless I'm misunderstanding what you're trying to do with this function (or it's supposed to be single voxel only) I don't see how it would work. Am I wrong here? To me I think you want `np.any` with the correct axis specified.